schema push

JakeDawkins opened the pull request:
Related to #2

JakeDawkins committed:
WIP on skeleton for stash command

JakeDawkins committed:
very roughly stashes a .graphql file

JakeDawkins committed:
update error handling and messaging

EverlastingBugstopper commented on crates/rover-client/src/query/schema/stash.rs:
```diff
@@ -0,0 +1,78 @@
+pub fn run(
+    variables: stash_schema_mutation::Variables,
+    client: Client,
+) -> Result<(String, String), RoverClientError> {
```
Would probably be good to create a struct here that has semantic meaning. so like `Result<crate::StashResponse>` that might look like:

```rust
struct StashResponse {
  schema_hash: GenericArray<u8, sha2::Digest::OutputSize>, // this could also just be a `String` if it's easier, though I like to keep non-String types around if at all possible
  update_status: String // or, like, an enum of all the possible values
  ```

you could even make a higher level enum that has all of the different possible responses from apollo studio stuff (but this part is not super necessary)

```rust
#[non_exhaustive] // this means there could be other responses besides the ones outlined here
// basically forces you to handle the `_` or default case
enum ApolloServerResponse {
  SchemaStash,
  SchemaPublish,
}
```

JakeDawkins replied:
> this could also just be a `String` if it's easier, though I like to keep non-String types around if at all possible

Can you explain this or provide a link so I can read up? I'm not sure why that'd be beneficial

EverlastingBugstopper replied:
I don't have a link really, this is more of just a personal preference where if there is a more specific data type you can provide, it makes it easier to extend/modify behavior moving forward in addition to making it obvious what TYPE of hash it actually is (instead of a String which could be any type of hash).

You'd just call `.to_string()` where you would want to use it instead of calling `.to_string()` before storing it in the struct.

EverlastingBugstopper commented on crates/rover-client/src/query/schema/stash.rs:
```diff
@@ -0,0 +1,78 @@
+            Some(data) => data,
+            None => {
+                return Err(RoverClientError::ResponseError {
+                    msg: "Error fetching schema. Check your API key & graph id".to_string(),
```
would be great if we could figure out which piece of info is incorrect or missing here (or if the response problem was a client disconnect they should just retry)

EverlastingBugstopper commented on src/command/schema/stash.rs:
```diff
@@ -0,0 +1,74 @@
+    schema_path: String,
+    /// The variant of the request graph from Apollo Studio
+    #[structopt(long, default_value = "current")]
+    variant: String,
```
what are the possible values here? would it be better to have this be an `enum`?

JakeDawkins replied:
Variants are user defined strings like `prod`, `current` (default), `dev`, `staging`, etc so they can be anything :)

EverlastingBugstopper commented on src/command/schema/stash.rs:
```diff
@@ -0,0 +1,74 @@
+pub struct Stash {
+    /// where to find the schema. .graphql, .json or uri
+    #[structopt(name = "SCHEMA_PATH")]
+    schema_path: String,
```
```suggestion
    schema_path: std::path::Path,
```

JakeDawkins replied:
Path won't work for me, but `PathBuf` does. I can't seem to get it to error before `run` though. Everything just gets caught in the execution

EverlastingBugstopper replied:
`PathBuf` is what I should have suggested here 

> I can't seem to get it to error before run though. Everything just gets caught in the execution

could you elaborate on this one? i'm not quite following 😄

EverlastingBugstopper commented on src/command/schema/stash.rs:
```diff
@@ -0,0 +1,74 @@
+                    "https://graphql.api.apollographql.com/api/graphql".to_string(),
+                );
+
+                let path = Path::new(&self.schema_path);
```
this will panic if it doesnt work which is why i recommended specifying this type in `StructOpt` so that error would show up at the arg parser step and not the run step

EverlastingBugstopper commented on src/command/schema/stash.rs:
```diff
@@ -0,0 +1,74 @@
+                );
+
+                let path = Path::new(&self.schema_path);
+                let contents = std::fs::read_to_string(path);
```
i'd recommend checking if the file exists here and if it doesn't returning an error before just trying to read the file to a string because it will panic if it doesn't exist

EverlastingBugstopper reviewed (commented):
looking good! haven't checked it out to test it, would love to see a demo application with this and the get command working in tandem 😄 

also likely gonna want some tests for this stuff (and it looks like `cargo clippy` isn't happy)

JakeDawkins committed:
update stash.run response to use response struct

JakeDawkins committed:
refactor stash::run

- move the deep data fetching to its own util function
- remove some nested matches

JakeDawkins committed:
Refactor fetch command

- use Errors where possible instead of panicking
- move file loading logic to util function
- use PathBuf instead of String for path arg

JakeDawkins committed:
fmt

JakeDawkins committed:
Add ability to push partial schemas from .graphql

JakeDawkins committed:
rename stash to push

JakeDawkins committed:
Restructure & add tests

- move partial schema pushes to /service/push
- changed command structure to use `service push` for partial schemas
- added tests for both rover-client `push` modules

JakeDawkins committed:
add more tests

JakeDawkins committed:
fix error names

JakeDawkins committed:
lint

EverlastingBugstopper commented on src/command/schema/push.rs:
```diff
@@ -0,0 +1,70 @@
+    #[structopt(long)]
+    graph: String,
+    #[structopt(long, default_value = "default")]
+    profile: String,
```
```suggestion
    #[structopt(long = "profile", default_value = "default")]
    profile_name: String,
```

EverlastingBugstopper commented on src/command/schema/push.rs:
```diff
@@ -0,0 +1,70 @@
+    #[structopt(long, default_value = "current")]
+    variant: String,
+    #[structopt(long)]
+    graph: String,
```
```suggestion
    #[structopt(long)]
    graph_name: String,
```

this should be `graph_name` to match what is in `schema get`.

i'm also guessing that it should be a required positional argument since you kinda need to know what graph you're pushing a schema to (unless not specifying a graph is a valid operation, in which case this is fine)

EverlastingBugstopper commented on src/command/schema/push.rs:
```diff
@@ -0,0 +1,70 @@
+
+    #[test]
+    fn get_schema_from_file_path_loads() {
+        // todo @jake -- add test for this after merging with avery's work
```
what is blocking this?

JakeDawkins replied:
It was the temp file stuff you had written for one of your PRs

EverlastingBugstopper replied:
ooh gotcha - i actually removed that in favor of using `assert_fs`, which is now in `main`. here's a few examples of using that library

https://github.com/apollographql/rover/blob/7b2dab780fb9c728f16730d751a5786961e6e150/src/command/config/api_key.rs#L58
https://github.com/apollographql/rover/pull/52/files#diff-6fcf03b28a7c3fdfc4ed29ed75e7fc1de9689f4e391b6443e5bd7de817c1d7e9R72

EverlastingBugstopper commented on src/command/schema/push.rs:
```diff
@@ -0,0 +1,70 @@
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+pub struct Push {
```
have we decided to omit federated graphs for now?

JakeDawkins replied:
Nope! That's handled in the `service push` command instead of here :)

EverlastingBugstopper commented on src/command/schema/push.rs:
```diff
@@ -0,0 +1,70 @@
+#[derive(Debug, StructOpt)]
+pub struct Push {
+    /// where to find the schema. .graphql, .json or uri
+    #[structopt(name = "SCHEMA_PATH", parse(from_os_str))]
```
does `from_os_str` handle URIs? i'm wondering if perhaps we should make this a bit more clear with some clap/structopt features that allow you to express this a bit differently (<PathBuf | Uri>) in the usage help text might be nice

EverlastingBugstopper reviewed (changes requested):
coming together!

JakeDawkins committed:
update arg names and add docs

JakeDawkins committed:
update SCHEMA_PATH arg to rm URI option

JakeDawkins committed:
add schema file loading tests

JakeDawkins committed:
Merge branch 'main' into jake/stash

JakeDawkins committed:
add command serialization

JakeDawkins committed:
skip serializing on PII

JakeDawkins committed:
merge main

JakeDawkins committed:
change log:: to tracing::

JakeDawkins committed:
change schema printing to println

JakeDawkins committed:
Merge branch 'main' into jake/stash
